lib/puppet/provider/reboot/windows.rb

```diff
@@ -66,19 +63,9 @@ def reboot
   end
 
   def async_shutdown(shutdown_cmd)
-    if Puppet[:debug]
-      $stderr.puts(shutdown_cmd)
-    end
-
-    # execute a ruby process to shutdown after puppet exits
-    watcher = File.join(File.dirname(__FILE__), 'windows', 'watcher.rb')
-    if not File.exists?(watcher)
-      raise ArgumentError, "The watcher program #{watcher} does not exist"
-    end
-
-    Puppet.debug("Launching 'ruby.exe #{watcher}'")
-    pid = Process.spawn("ruby.exe '#{watcher}' #{Process.pid} #{@resource[:catalog_apply_timeout]} '#{shutdown_cmd}'")
-    Puppet.debug("Launched process #{pid}")
+    Puppet.debug("Adding #{shutdown_cmd} to ruby's at_exit handler")
+    pid = Process.pid
+    at_exit { system shutdown_cmd if Process.pid == pid }
```

Comment:
That makes sense for posix, but ruby on windows doesn't implement `Kernel.fork`.